File: project-{topic}-advanced/src/backend/controllers/productController.js
```javascript
const { StatusCodes } = require('http-status-codes');
const Product = require('../models/Product');
const AppError = require('../utils/appError');
const catchAsync = require('../utils/catchAsync');
const productCache = require('../middleware/cacheMiddleware').productCache;

/**
 * @swagger
 * tags:
 *   name: Products
 *   description: Product management operations
 */

/**
 * @swagger
 * /products:
 *   get:
 *     summary: Get all products
 *     tags: [Products]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: query
 *         name: name
 *         schema:
 *           type: string
 *         description: Filter products by name
 *       - in: query
 *         name: category
 *         schema:
 *           type: string
 *         description: Filter products by category
 *       - in: query
 *         name: minPrice
 *         schema:
 *           type: number
 *         description: Filter products by minimum price
 *       - in: query
 *         name: maxPrice
 *         schema:
 *           type: number
 *         description: Filter products by maximum price
 *       - in: query
 *         name: sortBy
 *         schema:
 *           type: string
 *           enum: [name, price, createdAt]
 *         description: Field to sort by
 *       - in: query
 *         name: order
 *         schema:
 *           type: string
 *           enum: [ASC, DESC]
 *         default: ASC
 *         description: Sort order
 *       - in: query
 *         name: page
 *         schema:
 *           type: integer
 *           minimum: 1
 *         default: 1
 *         description: Page number
 *       - in: query
 *         name: limit
 *         schema:
 *           type: integer
 *           minimum: 1
 *           maximum: 100
 *         default: 10
 *         description: Number of products per page
 *     responses:
 *       200:
 *         description: List of products
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 status:
 *                   type: string
 *                   example: success
 *                 results:
 *                   type: integer
 *                   example: 1
 *                 data:
 *                   type: array
 *                   items:
 *                     $ref: '#/components/schemas/Product'
 *       401:
 *         $ref: '#/components/schemas/Error'
 */
exports.getAllProducts = catchAsync(async (req, res) => {
    const { name, category, minPrice, maxPrice, sortBy, order, page = 1, limit = 10 } = req.query;
    const queryOptions = { where: {} };

    // Filtering
    if (name) queryOptions.where.name = { [Product.sequelize.Op.iLike]: `%${name}%` };
    if (category) queryOptions.where.category = category;
    if (minPrice) queryOptions.where.price = { [Product.sequelize.Op.gte]: parseFloat(minPrice) };
    if (maxPrice) {
        queryOptions.where.price = {
            ...queryOptions.where.price,
            [Product.sequelize.Op.lte]: parseFloat(maxPrice),
        };
    }

    // Sorting
    const validSortFields = ['name', 'price', 'createdAt'];
    const sortField = validSortFields.includes(sortBy) ? sortBy : 'createdAt';
    const sortOrder = ['ASC', 'DESC'].includes(order?.toUpperCase()) ? order.toUpperCase() : 'DESC';
    queryOptions.order = [[sortField, sortOrder]];

    // Pagination
    const offset = (parseInt(page) - 1) * parseInt(limit);
    queryOptions.limit = parseInt(limit);
    queryOptions.offset = offset;

    const products = await Product.findAndCountAll(queryOptions);

    res.status(StatusCodes.OK).json({
        status: 'success',
        results: products.rows.length,
        total: products.count,
        page: parseInt(page),
        limit: parseInt(limit),
        data: products.rows,
    });
});

/**
 * @swagger
 * /products/{id}:
 *   get:
 *     summary: Get a product by ID
 *     tags: [Products]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *           format: uuid
 *         description: ID of the product to retrieve
 *     responses:
 *       200:
 *         description: Product details
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 status:
 *                   type: string
 *                   example: success
 *                 data:
 *                   $ref: '#/components/schemas/Product'
 *       401:
 *         $ref: '#/components/schemas/Error'
 *       404:
 *         $ref: '#/components/schemas/Error'
 */
exports.getProductById = catchAsync(async (req, res, next) => {
    const { id } = req.params;
    const product = await Product.findByPk(id);

    if (!product) {
        return next(new AppError('Product not found with that ID', StatusCodes.NOT_FOUND));
    }

    res.status(StatusCodes.OK).json({
        status: 'success',
        data: product,
    });
});

/**
 * @swagger
 * /products:
 *   post:
 *     summary: Create a new product (Admin only)
 *     tags: [Products]
 *     security:
 *       - bearerAuth: []
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             required:
 *               - name
 *               - description
 *               - price
 *               - quantity
 *             properties:
 *               name:
 *                 type: string
 *                 example: Smartphone X
 *               description:
 *                 type: string
 *                 example: A powerful smartphone with an amazing camera.
 *               price:
 *                 type: number
 *                 format: float
 *                 example: 799.99
 *               quantity:
 *                 type: integer
 *                 example: 100
 *               category:
 *                 type: string
 *                 example: Electronics
 *               imageUrl:
 *                 type: string
 *                 format: url
 *                 example: https://example.com/smartphone.jpg
 *     responses:
 *       201:
 *         description: Product created successfully
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 status:
 *                   type: string
 *                   example: success
 *                 data:
 *                   $ref: '#/components/schemas/Product'
 *       400:
 *         $ref: '#/components/schemas/Error'
 *       401:
 *         $ref: '#/components/schemas/Error'
 *       403:
 *         description: Forbidden - only administrators can create products
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/Error'
 */
exports.createProduct = catchAsync(async (req, res, next) => {
    // Clear cache for products after creation
    productCache.del('all_products');

    const newProduct = await Product.create(req.body);

    res.status(StatusCodes.CREATED).json({
        status: 'success',
        data: newProduct,
    });
});

/**
 * @swagger
 * /products/{id}:
 *   patch:
 *     summary: Update an existing product (Admin only)
 *     tags: [Products]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *           format: uuid
 *         description: ID of the product to update
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             properties:
 *               name:
 *                 type: string
 *                 example: Smartphone X Pro
 *               price:
 *                 type: number
 *                 format: float
 *                 example: 849.99
 *               quantity:
 *                 type: integer
 *                 example: 95
 *     responses:
 *       200:
 *         description: Product updated successfully
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 status:
 *                   type: string
 *                   example: success
 *                 data:
 *                   $ref: '#/components/schemas/Product'
 *       400:
 *         $ref: '#/components/schemas/Error'
 *       401:
 *         $ref: '#/components/schemas/Error'
 *       403:
 *         description: Forbidden - only administrators can update products
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/Error'
 *       404:
 *         $ref: '#/components/schemas/Error'
 */
exports.updateProduct = catchAsync(async (req, res, next) => {
    const { id } = req.params;
    const [updatedRowsCount, updatedProducts] = await Product.update(req.body, {
        where: { id },
        returning: true, // Return the updated object
    });

    if (updatedRowsCount === 0) {
        return next(new AppError('Product not found with that ID', StatusCodes.NOT_FOUND));
    }

    // Clear cache for specific product and all products
    productCache.del(`product_${id}`);
    productCache.del('all_products');

    res.status(StatusCodes.OK).json({
        status: 'success',
        data: updatedProducts[0],
    });
});

/**
 * @swagger
 * /products/{id}:
 *   delete:
 *     summary: Delete a product (Admin only)
 *     tags: [Products]
 *     security:
 *       - bearerAuth: []
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *           format: uuid
 *         description: ID of the product to delete
 *     responses:
 *       204:
 *         description: Product deleted successfully (No Content)
 *       401:
 *         $ref: '#/components/schemas/Error'
 *       403:
 *         description: Forbidden - only administrators can delete products
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/Error'
 *       404:
 *         $ref: '#/components/schemas/Error'
 */
exports.deleteProduct = catchAsync(async (req, res, next) => {
    const { id } = req.params;
    const deletedRowCount = await Product.destroy({ where: { id } });

    if (deletedRowCount === 0) {
        return next(new AppError('Product not found with that ID', StatusCodes.NOT_FOUND));
    }

    // Clear cache for specific product and all products
    productCache.del(`product_${id}`);
    productCache.del('all_products');

    res.status(StatusCodes.NO_CONTENT).json({
        status: 'success',
        data: null,
    });
});
```

File: project-{topic}-plus/src/main/java/com/alx/pm/controller/ProductController.java
```java
package com.alx.pm.controller;

import com.alx.pm.dto.ProductDTO;
import com.alx.pm.service.ProductService;
import io.github.bucket4j.Bucket;
import io.github.bucket4j.ConsumptionProbe;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/products")
@RequiredArgsConstructor
@Tag(name = "Product Management", description = "API for managing products")
public class ProductController {

    private final ProductService productService;
    private final Bucket rateLimitingBucket; // Injected rate limiting bucket

    @Operation(summary = "Get all products with pagination and filtering",
               responses = @ApiResponse(responseCode = "200", description = "List of products retrieved successfully"))
    @GetMapping
    public ResponseEntity<Page<ProductDTO>> getAllProducts(
            @RequestParam(defaultValue = "0") @Parameter(description = "Page number (0-indexed)") int page,
            @RequestParam(defaultValue = "10") @Parameter(description = "Number of items per page") int size,
            @RequestParam(defaultValue = "name,asc") @Parameter(description = "Sort by field and direction (e.g., 'name,asc')") String[] sort,
            @RequestParam(required = false) @Parameter(description = "Filter by product name") String name,
            @RequestParam(required = false) @Parameter(description = "Filter by product category ID") Long categoryId,
            @RequestParam(required = false) @Parameter(description = "Filter by minimum price") Double minPrice,
            @RequestParam(required = false) @Parameter(description = "Filter by maximum price") Double maxPrice
    ) {
        // Basic rate limiting check
        ConsumptionProbe probe = rateLimitingBucket.tryConsumeAndReturnRemaining(1);
        if (!probe.isConsumed()) {
            return ResponseEntity.status(HttpStatus.TOO_MANY_REQUESTS)
                    .header("X-Rate-Limit-Retry-After-Seconds", String.valueOf(probe.getNanosToWaitForRefill() / 1_000_000_000))
                    .build();
        }

        Sort sorting = Sort.by(Sort.Direction.fromString(sort[1]), sort[0]);
        PageRequest pageable = PageRequest.of(page, size, sorting);
        Page<ProductDTO> products = productService.getAllProducts(pageable, name, categoryId, minPrice, maxPrice);
        return ResponseEntity.ok(products);
    }

    @Operation(summary = "Get a product by ID",
               responses = {
                   @ApiResponse(responseCode = "200", description = "Product found"),
                   @ApiResponse(responseCode = "404", description = "Product not found")
               })
    @GetMapping("/{id}")
    public ResponseEntity<ProductDTO> getProductById(@PathVariable Long id) {
        ProductDTO product = productService.getProductById(id);
        return ResponseEntity.ok(product);
    }

    @Operation(summary = "Create a new product",
               responses = {
                   @ApiResponse(responseCode = "201", description = "Product created successfully"),
                   @ApiResponse(responseCode = "400", description = "Invalid input")
               })
    @PreAuthorize("hasRole('ADMIN') or hasRole('MANAGER')") // ADMINs and MANAGERS can create products
    @PostMapping
    public ResponseEntity<ProductDTO> createProduct(@Valid @RequestBody ProductDTO productDTO) {
        ProductDTO createdProduct = productService.createProduct(productDTO);
        return new ResponseEntity<>(createdProduct, HttpStatus.CREATED);
    }

    @Operation(summary = "Update an existing product",
               responses = {
                   @ApiResponse(responseCode = "200", description = "Product updated successfully"),
                   @ApiResponse(responseCode = "400", description = "Invalid input"),
                   @ApiResponse(responseCode = "404", description = "Product not found")
               })
    @PreAuthorize("hasRole('ADMIN') or hasRole('MANAGER')") // ADMINs and MANAGERS can update products
    @PutMapping("/{id}")
    public ResponseEntity<ProductDTO> updateProduct(@PathVariable Long id, @Valid @RequestBody ProductDTO productDTO) {
        ProductDTO updatedProduct = productService.updateProduct(id, productDTO);
        return ResponseEntity.ok(updatedProduct);
    }

    @Operation(summary = "Delete a product by ID",
               responses = {
                   @ApiResponse(responseCode = "204", description = "Product deleted successfully"),
                   @ApiResponse(responseCode = "404", description = "Product not found")
               })
    @PreAuthorize("hasRole('ADMIN')") // Only ADMINs can delete products
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteProduct(@PathVariable Long id) {
        productService.deleteProduct(id);
        return ResponseEntity.noContent().build();
    }
}
```